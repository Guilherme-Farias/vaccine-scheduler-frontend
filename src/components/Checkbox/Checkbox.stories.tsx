import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import Checkbox from '.';

export default {
  title: 'Form/Checkbox',
  component: Checkbox,
  argTypes: {
    label: {
      type: 'string',
      description: 'Checkbox text',
    },
    fillColor: {
      description: 'Change the input color',
      defaultValue: 'primary',
      control: {
        type: 'select',
        options: [
          'primary',
          'secondary',
          'tertiary',
          'green',
          'red',
          'yellow',
          'blue',
        ],
      },
    },
    disabled: {
      description: 'Change the state of the button',
      control: {
        type: 'boolean',
      },
    },
    checked: {
      type: 'boolean',
      description: 'Report status',
    },
  },
} as ComponentMeta<typeof Checkbox>;

const Template: ComponentStory<typeof Checkbox> = args => (
  <div style={{ display: 'grid', gap: '10px' }}>
    <Checkbox {...args} name="category_1" label="Action" />
    <Checkbox {...args} name="category_2" label="Adventure" />
    <Checkbox {...args} name="category_3" label="Strategy" />
  </div>
);

export const Default = Template.bind({});
Default.args = {};
