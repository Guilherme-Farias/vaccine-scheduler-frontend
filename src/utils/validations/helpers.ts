import Joi from 'joi';
import { IParamErrors } from 'types/global';

export function getFieldErrors(objError: Joi.ValidationResult) {
  const errors: IParamErrors = {};

  if (objError.error) {
    objError.error.details.forEach(err => {
      errors[err.path.join('.')] = err.message;
    });
  }

  return errors;
}
