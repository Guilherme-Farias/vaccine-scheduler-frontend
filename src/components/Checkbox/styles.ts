import styled, { css, DefaultTheme } from 'styled-components/macro';
import { darken } from 'polished';

import { FormGroup, FormLabel } from 'components/FormControls';
import { CheckboxProps } from './Checkbox';

type InputProps = Pick<CheckboxProps, 'fillColor'>;

export const Input = styled.input<InputProps>`
  ${({ theme, fillColor }) => css`
    cursor: pointer;
    appearance: none;
    display: flex;
    align-items: center;
    justify-content: center;
    width: var(--checkbox-size);
    min-width: var(--checkbox-size);
    height: var(--checkbox-size);
    min-height: var(--checkbox-size);
    border: 2px solid ${theme.colors['gray-600']};
    border-radius: 2px;
    position: relative;
    outline: none;
    transition: background-color 0.15s ease-in-out,
      border-color 0.15s ease-in-out, box-shadow 0.15s ease-in-out;

    &:before {
      content: '';
      width: 7px;
      height: 10px;
      border: 2.5px solid ${theme.colors.white};
      border-top: 0;
      border-left: 0;
      transform: rotate(45deg);
      position: absolute;
      top: 1px;
      opacity: 0;
      transition: 0.15s ease-in-out;
    }

    &:focus {
      box-shadow: ${theme.colors.blue} 0px 0px 0px 1px;
    }

    &:hover {
      border-color: ${theme.colors.black};
      transition: 0.15s ease-in-out;
    }

    &:checked {
      border-color: ${theme.colors[fillColor!]};
      background: ${theme.colors[fillColor!]};
      &:before {
        opacity: 1;
      }
    }
  `}
`;

export const StyledFormLabel = styled(FormLabel)`
  ${({ theme }) => css`
    // remove base FormLabel margin
    margin-bottom: revert;
    line-height: var(--checkbox-size);
    padding-left: ${theme.spacings.xs};
  `}
`;

const styledFormGroupModifiers = {
  isDisabled: (theme: DefaultTheme) => css`
    &,
    * {
      cursor: not-allowed;
    }

    ${Input} {
      border-color: ${theme.colors['gray-500']};
      background-color: ${theme.colors['gray-300']};
      &:checked {
        background: ${darken(0.2, theme.colors.primary)};
        border-color: ${darken(0.2, theme.colors.primary)};
      }
    }
  `,
};

export const StyledFormGroup = styled(FormGroup)`
  ${({ theme, isDisabled }) => css`
    display: flex;
    align-items: center;
    --checkbox-size: 20px;

    ${!!isDisabled && styledFormGroupModifiers.isDisabled(theme)};
  `}
`;
