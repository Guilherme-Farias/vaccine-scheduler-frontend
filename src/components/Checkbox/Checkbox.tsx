import React, { InputHTMLAttributes } from 'react';
import theme from 'styles/theme';
import * as S from './styles';

type ThemeColors = typeof theme.colors;

export type CheckboxColors = keyof Pick<
  ThemeColors,
  'primary' | 'secondary' | 'tertiary' | 'green' | 'red' | 'yellow' | 'blue'
>;

export type CheckboxProps = {
  label?: string;
  fillColor?: CheckboxColors;
} & InputHTMLAttributes<HTMLInputElement>;

const Checkbox: React.FC<CheckboxProps> = ({
  name,
  disabled,
  label = '',
  style,
  className,
  fillColor = 'primary',
  ...props
}: CheckboxProps) => (
  <S.StyledFormGroup isDisabled={disabled} className={className} style={style}>
    <S.Input
      name={name}
      type="checkbox"
      fillColor={fillColor}
      disabled={disabled}
      {...(label ? { id: name } : {})}
      {...props}
    />
    {!!label && (
      <S.StyledFormLabel htmlFor={name} isDisabled={disabled}>
        {label}
      </S.StyledFormLabel>
    )}
  </S.StyledFormGroup>
);

export default Checkbox;
