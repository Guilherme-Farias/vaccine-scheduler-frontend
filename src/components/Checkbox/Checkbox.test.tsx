import { render, screen, userEvent } from 'utils/tests/helpers';

import Checkbox from '.';

describe('<Checkbox />', () => {
  it('should render with label', () => {
    const { container } = render(
      <Checkbox name="checkbox_name" label="checkbox label" />,
    );
    expect(screen.getByRole('checkbox')).toBeInTheDocument();
    expect(screen.getByLabelText(/checkbox label/i)).toBeInTheDocument();
    expect(
      screen.getByText(/checkbox label/i).closest('label'),
    ).toHaveAttribute('for', 'checkbox_name');

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render without label', () => {
    render(<Checkbox />);

    expect(screen.queryByLabelText('Checkbox')).not.toBeInTheDocument();
  });

  it('should changes its value when checking', async () => {
    render(<Checkbox />);
    const checkbox = screen.getByRole('checkbox');

    await userEvent.click(checkbox);
    expect(checkbox).toBeChecked();
  });

  it('should call onCheck with false if the Checkbox is already checked', async () => {
    render(<Checkbox />);
    const checkbox = screen.getByRole('checkbox');

    await userEvent.click(checkbox);
    expect(checkbox).toBeChecked();

    await userEvent.click(checkbox);
    expect(checkbox).not.toBeChecked();
  });

  it('should not changes its value when disabled', async () => {
    render(<Checkbox name="checkbox_name" label="checkbox label" disabled />);

    const checkbox = screen.getByRole('checkbox');
    expect(checkbox).toBeDisabled();

    await userEvent.click(checkbox);

    expect(checkbox).not.toBeChecked();
  });

  it('should be accessible with tab', async () => {
    render(<Checkbox name="checkbox_name" label="checkbox label" />);

    expect(document.body).toHaveFocus();

    await userEvent.tab();

    expect(screen.getByLabelText(/checkbox label/i)).toHaveFocus();
  });

  it('should not accessible by tab when disabled', async () => {
    render(<Checkbox name="checkbox_name" label="checkbox label" disabled />);

    const checkbox = screen.getByLabelText('checkbox label');
    expect(document.body).toHaveFocus();

    await userEvent.tab();
    expect(checkbox).not.toHaveFocus();
  });
});
