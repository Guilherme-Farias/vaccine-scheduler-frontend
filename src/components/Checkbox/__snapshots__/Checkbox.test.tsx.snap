// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> should render with label 1`] = `
.c2 {
  cursor: pointer;
  display: inline-block;
  color: #171923;
  font-size: 1rem;
  font-weight: 500;
  margin-bottom: 0.5rem;
}

.c1 {
  cursor: pointer;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: var(--checkbox-size);
  min-width: var(--checkbox-size);
  height: var(--checkbox-size);
  min-height: var(--checkbox-size);
  border: 2px solid #4b5563;
  border-radius: 2px;
  position: relative;
  outline: none;
  -webkit-transition: background-color 0.15s ease-in-out, border-color 0.15s ease-in-out,box-shadow 0.15s ease-in-out;
  transition: background-color 0.15s ease-in-out, border-color 0.15s ease-in-out,box-shadow 0.15s ease-in-out;
}

.c1:before {
  content: '';
  width: 7px;
  height: 10px;
  border: 2.5px solid #FFFFFF;
  border-top: 0;
  border-left: 0;
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
  position: absolute;
  top: 1px;
  opacity: 0;
  -webkit-transition: 0.15s ease-in-out;
  transition: 0.15s ease-in-out;
}

.c1:focus {
  box-shadow: #2563eb 0px 0px 0px 1px;
}

.c1:hover {
  border-color: #171923;
  -webkit-transition: 0.15s ease-in-out;
  transition: 0.15s ease-in-out;
}

.c1:checked {
  border-color: #D03E35;
  background: #D03E35;
}

.c1:checked:before {
  opacity: 1;
}

.c3 {
  margin-bottom: revert;
  line-height: var(--checkbox-size);
  padding-left: 0.5rem;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  --checkbox-size: 20px;
}

<div
  class=" c0"
  role="group"
>
  <input
    class="c1"
    id="checkbox_name"
    name="checkbox_name"
    type="checkbox"
  />
  <label
    class="c2 c3"
    for="checkbox_name"
  >
    checkbox label
  </label>
</div>
`;
